{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "load obo file C:\\Users\\Daniel\\.pyppi\\go.obo\n",
      "C:\\Users\\Daniel\\.pyppi\\go.obo: fmt(1.2) rel(2017-07-14) 48,971 GO Terms\n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "\"\"\"\n",
    "This script runs the bootstrap kfold validation experiments as used in\n",
    "the publication.\n",
    "\n",
    "Usage:\n",
    "  validation.py [--interpro] [--pfam] [--mf] [--cc] [--bp]\n",
    "             [--use_cache] [--induce] [--verbose]\n",
    "             [--model=M] [--n_jobs=J] [--n_splits=S] [--n_iterations=I]\n",
    "             [--directory=DIR]\n",
    "  validation.py -h | --help\n",
    "\n",
    "Options:\n",
    "  -h --help     Show this screen.\n",
    "  --interpro    Use interpro domains in features.\n",
    "  --pfam        Use Pfam domains in features.\n",
    "  --mf          Use Molecular Function Gene Ontology in features.\n",
    "  --cc          Use Cellular Compartment Gene Ontology in features.\n",
    "  --bp          Use Biological Process Gene Ontology in features.\n",
    "  --induce      Use ULCA inducer over Gene Ontology.\n",
    "  --verbose     Print intermediate output for debugging.\n",
    "  --use_cache   Use cached features if available.\n",
    "  --model=M         A binary classifier from Scikit-Learn implementing fit,\n",
    "                    predict and predict_proba [default: LogisticRegression]\n",
    "  --n_jobs=J        Number of processes to run in parallel [default: 1]\n",
    "  --n_splits=S      Number of cross-validation splits [default: 5]\n",
    "  --n_iterations=I  Number of bootstrap iterations [default: 5]\n",
    "  --directory=DIR   Output directory [default: ./results/]\n",
    "\"\"\"\n",
    "\n",
    "import json\n",
    "import logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from operator import itemgetter\n",
    "from collections import Counter\n",
    "from datetime import datetime\n",
    "\n",
    "from pyppi.base import parse_args, su_make_dir\n",
    "from pyppi.data import load_network_from_path, load_ptm_labels\n",
    "from pyppi.data import testing_network_path, training_network_path\n",
    "from pyppi.data import load_go_dag, ipr_name_map, pfam_name_map\n",
    "\n",
    "from pyppi.models.binary_relevance import BinaryRelevance\n",
    "from pyppi.models import make_classifier\n",
    "from pyppi.model_selection.scoring import MultilabelScorer, Statistics\n",
    "from pyppi.model_selection.experiment import KFoldExperiment, Bootstrap\n",
    "from pyppi.model_selection.sampling import IterativeStratifiedKFold\n",
    "\n",
    "from pyppi.data_mining.features import AnnotationExtractor\n",
    "from pyppi.data_mining.uniprot import UniProt, get_active_instance\n",
    "from pyppi.data_mining.tools import xy_from_interaction_frame\n",
    "\n",
    "from pyppi.data_mining.ontology import id_to_node\n",
    "\n",
    "from sklearn.base import clone\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.multiclass import OneVsRestClassifier\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "from sklearn.metrics import f1_score, precision_score, recall_score\n",
    "from sklearn.metrics import (\n",
    "    recall_score, make_scorer, \n",
    "    label_ranking_average_precision_score,\n",
    "    label_ranking_loss\n",
    ")\n",
    "\n",
    "from sklearn.datasets import make_multilabel_classification\n",
    "\n",
    "logging.basicConfig(\n",
    "    format='[%(asctime)s] %(levelname)s: %(message)s', \n",
    "    datefmt='%m-%d-%Y %I:%M:%S',\n",
    "    level=logging.DEBUG,\n",
    ")\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "args = {\n",
    "    'n_jobs': 1,\n",
    "    'n_splits': 5,\n",
    "    'n_iterations': 5,\n",
    "    'induce': True,\n",
    "    'verbose': True,\n",
    "    'selection': [\n",
    "        UniProt.data_types().GO_MF.value,\n",
    "        UniProt.data_types().GO_BP.value,\n",
    "        UniProt.data_types().GO_CC.value,\n",
    "        UniProt.data_types().INTERPRO.value,\n",
    "        UniProt.data_types().PFAM.value\n",
    "    ],\n",
    "    'model': 'LogisticRegression',\n",
    "    'use_cache': True,\n",
    "    'directory': './results/'\n",
    "}\n",
    "n_jobs = args['n_jobs']\n",
    "n_splits = args['n_splits']\n",
    "n_iter = args['n_iterations']\n",
    "induce = args['induce']\n",
    "verbose = args['verbose']\n",
    "selection = args['selection']\n",
    "model = args['model']\n",
    "use_feature_cache = args['use_cache']\n",
    "direc = args['directory']\n",
    "backend = 'multiprocessing'\n",
    "get_top_n = 25\n",
    "go_dag = None\n",
    "\n",
    "# Set up the folder for each experiment run named after the current time\n",
    "folder = datetime.now().strftime(\"val_%y-%m-%d_%H-%M\")\n",
    "direc = \"{}/{}/\".format(direc, folder)\n",
    "su_make_dir(direc)\n",
    "json.dump(\n",
    "    args, fp=open(\"{}/settings.json\".format(direc), 'w'),\n",
    "    indent=4, sort_keys=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:42:42] INFO: Loading training and testing data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "load obo file C:\\Users\\Daniel\\.pyppi\\go.obo\n",
      "C:\\Users\\Daniel\\.pyppi\\go.obo: fmt(1.2) rel(2017-07-14) 48,971 GO Terms\n",
      "First time loading on UniProt instance. Make take a few moments\n",
      "Warning: Loading dat files, may take a few minutes.\n"
     ]
    }
   ],
   "source": [
    "logging.info(\"Loading training and testing data.\")\n",
    "if not go_dag:\n",
    "    go_dag = load_go_dag()\n",
    "ipr_map = ipr_name_map(lowercase_keys=False)\n",
    "pfam_map = pfam_name_map(lowercase_keys=False)\n",
    "uniprot = get_active_instance(\n",
    "    verbose=verbose,\n",
    "    sprot_cache=None,\n",
    "    trembl_cache=None\n",
    ")\n",
    "data_types = UniProt.data_types()\n",
    "labels = load_ptm_labels()\n",
    "annotation_ex = AnnotationExtractor(\n",
    "    induce=induce,\n",
    "    selection=selection,\n",
    "    n_jobs=n_jobs,\n",
    "    verbose=verbose,\n",
    "    cache=use_feature_cache,\n",
    "    backend='multiprocessing'\n",
    ")\n",
    "training = load_network_from_path(training_network_path)\n",
    "testing = load_network_from_path(testing_network_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:43:34] INFO: Preparing training and testing data.\n",
      "[08-17-2017 03:43:34] INFO: Computing class distributions.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding new PPIs...\n",
      "Stringing selected features for each PPI...\n",
      "Finding new PPIs...\n",
      "Stringing selected features for each PPI...\n"
     ]
    }
   ],
   "source": [
    "# Get the features into X, and multilabel y indicator format\n",
    "logging.info(\"Preparing training and testing data.\")\n",
    "mlb = MultiLabelBinarizer(classes=labels)\n",
    "X_train_ppis, y_train = xy_from_interaction_frame(training)\n",
    "X_test_ppis, y_test = xy_from_interaction_frame(testing)\n",
    "mlb.fit(y_train)\n",
    "\n",
    "logging.info(\"Computing class distributions.\")\n",
    "json.dump(\n",
    "    Counter([l for ls in y_train for l in ls]), \n",
    "    fp=open(\"{}/training_distribution.json\".format(direc), 'w'),\n",
    "    indent=4, sort_keys=True\n",
    ")\n",
    "json.dump(\n",
    "    Counter([l for ls in y_test for l in ls]), \n",
    "    fp=open(\"{}/testing_distribution.json\".format(direc), 'w'),\n",
    "    indent=4, sort_keys=True\n",
    ")\n",
    "\n",
    "X_train = annotation_ex.transform(X_train_ppis)\n",
    "X_test = annotation_ex.transform(X_test_ppis)\n",
    "y_train = mlb.transform(y_train)\n",
    "y_test = mlb.transform(y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def specificity(y_true, y_pred):\n",
    "    y_true = np.asarray(list(y_true))\n",
    "    y_pred = np.asarray(list(y_pred))\n",
    "    tn = np.sum(y_true == 0)\n",
    "    fp = np.sum([True for (a, b) in zip(y_true, y_pred) if a == 0 and b == 1])\n",
    "    if (tn + fp) == 0:\n",
    "        return np.NaN\n",
    "    return tn / (tn + fp)\n",
    "\n",
    "\n",
    "def get_coefs(clf):\n",
    "    \"\"\"\n",
    "    Return the feature weightings for each estimator. If estimator is a\n",
    "    pipeline, then it assumes the last step is the estimator.\n",
    "\n",
    "    :return: array-like, shape (n_classes_, n_features)\n",
    "    \"\"\"\n",
    "    def feature_imp(estimator):\n",
    "        if hasattr(estimator, 'steps'):\n",
    "            estimator = estimator.steps[-1][-1]\n",
    "        if hasattr(estimator, \"coef_\"):\n",
    "            return estimator.coef_\n",
    "        elif hasattr(estimator, \"coefs_\"):\n",
    "            return estimator.coefs_\n",
    "        elif hasattr(estimator, \"feature_importances_\"):\n",
    "            return estimator.feature_importances_\n",
    "        else:\n",
    "            raise AttributeError(\n",
    "                \"Estimator {} doesn't support \"\n",
    "                \"feature coefficients.\".format(type(estimator)))\n",
    "\n",
    "    return feature_imp(clf.best_estimator_)\n",
    "\n",
    "\n",
    "def top_n_features(n, clf, absolute=True, vectorizer=None):\n",
    "    \"\"\"\n",
    "    Return the top N features. If clf is a pipeline, then it assumes\n",
    "    the first step is the vectoriser holding the feature names.\n",
    "\n",
    "    :return: array like, shape (n_estimators, n).\n",
    "        Each element in a list is a tuple (feature_idx, weight).\n",
    "    \"\"\"\n",
    "    top_features = []\n",
    "    coefs = get_coefs(clf)[0]\n",
    "\n",
    "    if absolute:\n",
    "        coefs = abs(coefs)\n",
    "    if hasattr(clf, 'steps') and vectorizer is None:\n",
    "        vectorizer = clf.steps[0][-1]\n",
    "    idx_coefs = sorted(\n",
    "        enumerate(coefs), key=itemgetter(1), reverse=True\n",
    "    )[:n]\n",
    "    if vectorizer:\n",
    "        idx = [idx for (idx, w) in idx_coefs]\n",
    "        ws = [w for (idx, w) in idx_coefs]\n",
    "        names = np.asarray(vectorizer.get_feature_names())[idx]\n",
    "        return list(zip(names, ws))\n",
    "    else:\n",
    "        return idx_coefs\n",
    "    \n",
    "def get_term_name(go_dag, term):\n",
    "    if 'go' in term.lower():\n",
    "        term = term.replace(\"go\", \"GO:\")\n",
    "        return go_dag[term.upper()].name\n",
    "    elif 'ipr' in term.lower():\n",
    "        return ipr_map[term.upper()]\n",
    "    elif 'pf' in term.lower():\n",
    "        return pfam_map[term.upper()]\n",
    "    return None\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:45:44] INFO: Setting up preliminaries and the statistics arrays\n",
      "[08-17-2017 03:45:44] INFO: Found classes carboxylation, methylation, dephosphorylation, activation, sulfation, phosphorylation, deacetylation, prenylation, ubiquitination, proteolytic-cleavage, binding/association, acetylation, glycosylation, inhibition, state-change, dissociation, myristoylation, sumoylation\n"
     ]
    }
   ],
   "source": [
    "logging.info(\"Setting up preliminaries and the statistics arrays\")\n",
    "logging.info(\"Found classes {}\".format(', '.join(mlb.classes)))\n",
    "n_classes = len(mlb.classes)\n",
    "seeds = range(n_iter)\n",
    "top_features = {l:[] for l in labels}\n",
    "param_distribution = {\n",
    "    'C': np.arange(0.01, 20.01, step=0.01),\n",
    "    'penalty': ['l1', 'l2']\n",
    "}\n",
    "\n",
    "binary_scoring_funcs = [\n",
    "    ('Binary F1', f1_score) , \n",
    "    ('Precision', precision_score), \n",
    "    ('Recall', recall_score),\n",
    "    ('Specificity', recall_score)\n",
    "]\n",
    "multilabel_scores_funcs = [\n",
    "    ('Label Ranking Loss', label_ranking_loss), \n",
    "    ('Label Ranking Average Precision', label_ranking_average_precision_score), \n",
    "    ('Macro (weighted) F1', f1_score), \n",
    "    ('Macro (un-weighted) F1', f1_score)\n",
    "]\n",
    "n_scorers = len(binary_scoring_funcs)\n",
    "n_ml_scorers = len(multilabel_scores_funcs)\n",
    "\n",
    "# 2: position 0 is for validation, position 1 is for testing\n",
    "binary_statistics = np.zeros((n_iter, n_splits, n_classes, 2, n_scorers))\n",
    "multilabel_statistics = np.zeros((n_iter, n_splits, 2, n_ml_scorers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:45:45] INFO: Fitting bootstrap iteration 1.\n",
      "[08-17-2017 03:45:48] INFO: Fitting fold iteration 1.\n",
      "[08-17-2017 03:45:48] INFO: Fitting label carboxylation.\n",
      "[08-17-2017 03:45:48] INFO: Preparing data.\n",
      "[08-17-2017 03:45:58] INFO: Fitting classifier.\n",
      "[08-17-2017 03:46:01] INFO: Computing fold label binary performance.\n",
      "[08-17-2017 03:46:01] INFO: Computing top label features for fold.\n",
      "[08-17-2017 03:46:01] INFO: Computing fold mult-label performance.\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "y_true and y_score have different shape",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-6-774b2313c572>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     94\u001b[0m                 \u001b[0mscores_t\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my_test_f\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my_test_f_proba\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     95\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 96\u001b[1;33m                 \u001b[0mscores_v\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my_valid_f\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my_valid_f_pred\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     97\u001b[0m                 \u001b[0mscores_t\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my_test_f\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my_test_f_pred\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     98\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\envs\\pyppi\\lib\\site-packages\\scikit_learn-0.18.2-py3.5-win-amd64.egg\\sklearn\\metrics\\ranking.py\u001b[0m in \u001b[0;36mlabel_ranking_loss\u001b[1;34m(y_true, y_score, sample_weight)\u001b[0m\n\u001b[0;32m    728\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    729\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0my_true\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0my_score\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 730\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"y_true and y_score have different shape\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    731\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    732\u001b[0m     \u001b[0mn_samples\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_labels\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0my_true\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: y_true and y_score have different shape"
     ]
    }
   ],
   "source": [
    "for bs_iter in range(n_iter):\n",
    "    logging.info(\"Fitting bootstrap iteration {}.\".format(bs_iter + 1))\n",
    "    cv = IterativeStratifiedKFold(n_splits=n_splits, random_state=seeds[bs_iter])\n",
    "    \n",
    "    for fold_iter, (train_idx, validation_idx) in enumerate(cv.split(X_train, y_train)):\n",
    "        logging.info(\"Fitting fold iteration {}.\".format(fold_iter + 1))\n",
    "        y_valid_f_pred = []\n",
    "        y_test_f_pred = []\n",
    "        y_valid_f_proba = []\n",
    "        y_test_f_proba = []\n",
    "        \n",
    "        for label_idx, label in enumerate(labels):\n",
    "            logging.info(\"Fitting label {}.\".format(label))\n",
    "            \n",
    "            # Prepare all training and testing data\n",
    "            logging.info(\"Preparing data.\")\n",
    "            vectorizer = CountVectorizer(binary=False)\n",
    "            vectorizer.fit(X_train[train_idx])\n",
    "\n",
    "            X_train_l = vectorizer.transform(X_train[train_idx])\n",
    "            y_train_l = y_train[train_idx, label_idx]\n",
    "        \n",
    "            X_valid_l = vectorizer.transform(X_train[validation_idx])\n",
    "            y_valid_l = y_train[validation_idx, label_idx]\n",
    "\n",
    "            X_test_l = vectorizer.transform(X_test)\n",
    "            y_test_l = y_test[:, label_idx]\n",
    "            \n",
    "            # Build and fit classifier\n",
    "            logging.info(\"Fitting classifier.\")\n",
    "            clf = RandomizedSearchCV(\n",
    "                estimator=make_classifier(algorithm=model, random_state=0),\n",
    "                scoring='f1', cv=3, n_iter=1, n_jobs=n_jobs, refit=True,\n",
    "                random_state=0, param_distributions=param_distribution,\n",
    "            )\n",
    "            clf.fit(X_train_l, y_train_l)\n",
    "            \n",
    "            # Validation scores in binary and probability format\n",
    "            y_valid_l_pred = clf.predict(X_valid_l)\n",
    "            y_valid_l_proba = clf.predict_proba(X_valid_l)\n",
    "            \n",
    "            # Held-out testing scores in binary and probability format\n",
    "            y_test_l_pred = clf.predict(X_test_l)\n",
    "            y_test_l_proba = clf.predict_proba(X_test_l)\n",
    "            \n",
    "            # Store these per label results in a list which we will\n",
    "            # later use to stack into a multi-label array.\n",
    "            y_valid_f_pred.append([[x] for x in y_valid_l_pred])\n",
    "            y_valid_f_proba.append([[x[1]] for x in y_valid_l_proba])\n",
    "            \n",
    "            y_test_f_pred.append([[x] for x in y_test_l_pred])\n",
    "            y_test_f_proba.append([[x[1]] for x in y_test_l_proba])\n",
    "            \n",
    "            # Perform scoring on the validation set and the external testing set.\n",
    "            logging.info(\"Computing fold label binary performance.\")\n",
    "            for func_idx, (_, func) in enumerate(binary_scoring_funcs):\n",
    "                scores_v = func(y_valid_l, y_valid_l_pred, average='binary')\n",
    "                scores_t = func(y_test_l, y_test_l_pred, average='binary')\n",
    "                binary_statistics[bs_iter, fold_iter, label_idx, 0, func_idx] = scores_v\n",
    "                binary_statistics[bs_iter, fold_iter, label_idx, 1, func_idx] = scores_t\n",
    "                \n",
    "            logging.info(\"Computing top label features for fold.\")\n",
    "            # Get the top 20 features for this labels's run.\n",
    "            top_n = [\n",
    "                f for f, _ in \n",
    "                top_n_features(\n",
    "                    clf=clf, n=get_top_n, absolute=True, vectorizer=vectorizer)\n",
    "            ]\n",
    "            top_features[label].extend(top_n)\n",
    "            break\n",
    "        \n",
    "        logging.info(\"Computing fold mult-label performance.\")\n",
    "        # True scores in multi-label indicator format\n",
    "        y_valid_f = y_train[validation_idx]\n",
    "        y_test_f = y_test\n",
    "        \n",
    "        # Validation scores in multi-label indicator format\n",
    "        y_valid_f_pred = np.hstack(y_valid_f_pred)\n",
    "        y_valid_f_proba = np.hstack(y_valid_f_proba)\n",
    "        \n",
    "        # Testing scores in multi-label probability format\n",
    "        y_test_f_pred = np.hstack(y_test_f_pred)\n",
    "        y_test_f_proba = np.hstack(y_test_f_proba)\n",
    "        \n",
    "        for func_idx, (func_name, func) in enumerate(multilabel_scores_funcs):\n",
    "            if func_name == 'Macro (weighted) F1':\n",
    "                scores_v = func(y_valid_f, y_valid_f_pred, average='weighted')\n",
    "                scores_t = func(y_test_f, y_test_f_pred, average='weighted')\n",
    "            elif func_name == 'Macro (un-weighted) F1':\n",
    "                scores_v = func(y_valid_f, y_valid_f_pred, average='macro')\n",
    "                scores_t = func(y_test_f, y_test_f_pred, average='macro')\n",
    "            elif func_name == 'Label Ranking Average Precision':\n",
    "                scores_v = func(y_valid_f, y_valid_f_proba)\n",
    "                scores_t = func(y_test_f, y_test_f_proba)\n",
    "            else:\n",
    "                scores_v = func(y_valid_f, y_valid_f_pred)\n",
    "                scores_t = func(y_test_f, y_test_f_pred)\n",
    "                \n",
    "            multilabel_statistics[bs_iter, fold_iter, 0, func_idx] = scores_v\n",
    "            multilabel_statistics[bs_iter, fold_iter, 1, func_idx] = scores_t\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "logging.info(\"Writing statistics to file.\")\n",
    "func_names = [n for n, _ in binary_scoring_funcs]\n",
    "iterables = [range(n_iter), range(n_splits), range(n_classes), [\"validation\", \"holdout\"], func_names]\n",
    "names=['bootstrap iteration', 'fold iteration', 'labels', 'condition', 'score function']\n",
    "index = pd.MultiIndex.from_product(iterables, names=names)\n",
    "binary_df = pd.DataFrame(binary_statistics.ravel(), index=index)\n",
    "binary_df.to_csv('{}/{}'.format(folder, 'binary_stats.csv'), sep=',', index=False)\n",
    "\n",
    "func_names = [n for n, _ in multilabel_scores_funcs]\n",
    "iterables = [range(n_iter), range(n_splits), [\"validation\", \"holdout\"], func_names]\n",
    "names=['bootstrap iteration', 'fold iteration', 'condition', 'score function']\n",
    "index = pd.MultiIndex.from_product(iterables, names=names)\n",
    "multilabel_df = pd.DataFrame(multilabel_statistics.ravel(), index=index)\n",
    "multilabel_df.to_csv('{}/{}'.format(folder, 'multilabel_stats.csv'), sep=',', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:51:52] INFO: Computing top label features overall.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [list(['go0000018', 'regulation of DNA recombination'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [list(['go0000018', 'regulation of DNA recombination'])]\n",
      " [list(['ipr007782', 'VKG_COASE'])]\n",
      " [0]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [list(['go0000018', 'regulation of DNA recombination'])]\n",
      " [list(['ipr007782', 'VKG_COASE'])]\n",
      " [list(['pf05090', 'VITAMIN K-DEPENDENT GAMMA-CARBOXYLASE'])]\n",
      " [0]\n",
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [list(['go0000018', 'regulation of DNA recombination'])]\n",
      " [list(['ipr007782', 'VKG_COASE'])]\n",
      " [list(['pf05090', 'VITAMIN K-DEPENDENT GAMMA-CARBOXYLASE'])]\n",
      " [list(['go0005829', 'cytosol'])]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [0]]\n",
      "[[list(['go0005886', 'plasma membrane'])]\n",
      " [list(['go0000002', 'mitochondrial genome maintenance'])]\n",
      " [list(['go0000014', 'single-stranded DNA endodeoxyribonuclease activity'])]\n",
      " [list(['go0000019', 'regulation of mitotic recombination'])]\n",
      " [list(['go0005515', 'protein binding'])]\n",
      " [list(['go0008488', 'gamma-glutamyl carboxylase activity'])]\n",
      " [list(['go0000041', 'transition metal ion transport'])]\n",
      " [list(['go0003824', 'catalytic activity'])]\n",
      " [list(['go0000011', 'vacuole inheritance'])]\n",
      " [list(['ipr014710', 'RMLC-LIKE_JELLYROLL'])]\n",
      " [list(['ipr011020', 'HTTM'])]\n",
      " [list(['go0016831', 'carboxy-lyase activity'])]\n",
      " [list(['go0000035', 'acyl binding'])]\n",
      " [list(['go0000012', 'single strand break repair'])]\n",
      " [list(['go0043231', 'intracellular membrane-bounded organelle'])]\n",
      " [list(['go0018200', 'peptidyl-glutamic acid modification'])]\n",
      " [list(['go0016830', 'carbon-carbon lyase activity'])]\n",
      " [list(['ipr011051', 'RMLC_CUPIN'])]\n",
      " [list(['go0007165', 'signal transduction'])]\n",
      " [list(['go0046872', 'metal ion binding'])]\n",
      " [list(['go0000018', 'regulation of DNA recombination'])]\n",
      " [list(['ipr007782', 'VKG_COASE'])]\n",
      " [list(['pf05090', 'VITAMIN K-DEPENDENT GAMMA-CARBOXYLASE'])]\n",
      " [list(['go0005829', 'cytosol'])]\n",
      " [list(['go0017187', 'peptidyl-glutamic acid carboxylation'])]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08-17-2017 03:51:52] INFO: Writing top features to file.\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'val_17-08-17_15-42/top_features.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-14-16b20c96517a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     18\u001b[0m \u001b[0mlogging\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Writing top features to file.\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[0mtop_features_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 20\u001b[1;33m \u001b[0mtop_features_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'{}/{}'\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfolder\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'top_features.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msep\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m','\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\envs\\pyppi\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mto_csv\u001b[1;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, tupleize_cols, date_format, doublequote, escapechar, decimal)\u001b[0m\n\u001b[0;32m   1401\u001b[0m                                      \u001b[0mdoublequote\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdoublequote\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1402\u001b[0m                                      escapechar=escapechar, decimal=decimal)\n\u001b[1;32m-> 1403\u001b[1;33m         \u001b[0mformatter\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msave\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1404\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1405\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mpath_or_buf\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\envs\\pyppi\\lib\\site-packages\\pandas\\io\\formats\\format.py\u001b[0m in \u001b[0;36msave\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1569\u001b[0m             f, handles = _get_handle(self.path_or_buf, self.mode,\n\u001b[0;32m   1570\u001b[0m                                      \u001b[0mencoding\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1571\u001b[1;33m                                      compression=self.compression)\n\u001b[0m\u001b[0;32m   1572\u001b[0m             \u001b[0mclose\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1573\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\envs\\pyppi\\lib\\site-packages\\pandas\\io\\common.py\u001b[0m in \u001b[0;36m_get_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text)\u001b[0m\n\u001b[0;32m    383\u001b[0m         \u001b[1;32melif\u001b[0m \u001b[0mis_text\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    384\u001b[0m             \u001b[1;31m# Python 3 and no explicit encoding\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 385\u001b[1;33m             \u001b[0mf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath_or_buf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'replace'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    386\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    387\u001b[0m             \u001b[1;31m# Python 3 and binary mode\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'val_17-08-17_15-42/top_features.csv'"
     ]
    }
   ],
   "source": [
    "logging.info(\"Computing top label features overall.\")\n",
    "iterables = [mlb.classes, range(get_top_n), range(1)]\n",
    "names=['label', 'name', 'description']\n",
    "index = pd.MultiIndex.from_product(iterables, names=names)\n",
    "data = np.zeros(shape=(len(mlb.classes), get_top_n, 1), dtype='object')\n",
    "\n",
    "for i, (label, feature_ls) in enumerate([(l, top_features[l]) for l in mlb.classes]):\n",
    "    counts = Counter(feature_ls)\n",
    "    counts = sorted(counts.items(), key=itemgetter(1), reverse=True)\n",
    "    top_n_terms = [term for term, _ in counts][:get_top_n]\n",
    "    top_n_desc = [get_term_name(go_dag, t) for t in top_n_terms]\n",
    "    \n",
    "    for j, (term, desc) in enumerate(zip(top_n_terms, top_n_desc)):\n",
    "        data[i, j, 0] = [term, desc]\n",
    "        print(data[i])\n",
    "\n",
    "    \n",
    "logging.info(\"Writing top features to file.\")\n",
    "top_features_df = pd.DataFrame(data=data.ravel(), index=index)   \n",
    "top_features_df.to_csv('{}/{}'.format(folder, 'top_features.csv'), sep=',', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_features_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pyppi]",
   "language": "python",
   "name": "conda-env-pyppi-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
